onnx/test/cpp/data_propagation_test.cc

```diff
@@ -0,0 +1,156 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <iostream>
+#include "gtest/gtest.h"
+#include "onnx/checker.h"
+#include "onnx/defs/parser.h"
+#include "onnx/defs/schema.h"
+#include "onnx/defs/shape_inference.h"
+#include "onnx/onnx_pb.h"
+
+#include "onnx/shape_inference/implementation.h"
+
+using namespace ONNX_NAMESPACE::shape_inference;
+
+namespace ONNX_NAMESPACE {
+
+namespace Test {
+
+inline bool CompareShape(const TensorShapeProto* A, const TensorShapeProto* B) {
+    if (A == nullptr ||  B == nullptr) {
+        fail_check("The compared shapes should not be nullptr.");
+        return false;
+    }
+    if (A->dim_size() != B->dim_size()) {
+        fail_check("The compared sizes of dim are different.");
+        return false;
+    }
+    for (int i = 0; i < A->dim_size() ; ++i) {
+        if (A->dim(i).has_dim_value() != B->dim(i).has_dim_value() || 
+                A->dim(i).dim_value() != B->dim(i).dim_value()) {
```

Comment:
this check does not seem right... 
if A->dim(i).has_dim_value() is false then you should not fetch dim value... also why is there no check for dim_param
from the function name it seems like you are simply comparing the shapes but you are actually just checking the dim values